Change to be independent of the `yaml` and `file` package

masa-futa opened the pull request:
## Issue

- #138 

## Overview (Required)
yaml was used to convert the pubspecFile in the File system to yaml format. `loadYaml`
In order to maintain the previous behavior, we have implemented the actions performed by the yaml package manually
and have discontinued the yaml package.
In addition, because its use is limited, processing was limited to the bare minimum.


## Screenshot


<img src="https://github.com/user-attachments/assets/3f07b9c4-ba03-4388-8e79-afd4b0457c15" width="300" />

masa-futa commented:
![memo-badge](https://img.shields.io/badge/review-memo-lightgrey)

Changed `Yaml.parse` argument to `string` for easier testing.
Because of this, `MemoryFileSystem` is no longer necessary, so added support for removing it.

https://github.com/yumemi-inc/flutter-yumemi-lints/pull/149/commits/634c46288a723639b391051cdeb214df3552b88c

blendthink commented on packages/yumemi_lints/lib/src/models/yaml.dart:
```diff
@@ -0,0 +1,63 @@
+      while (stack.length > indent) {
+        stack.removeLast();
+        current = stack.isEmpty
+            ? result
+            : current['__parent__'] as Map<String, dynamic>;
+      }
+
+      if (trimmed.contains(':')) {
+        final parts = trimmed.split(':');
+        final key = parts[0].trim();
+        final value = parts.sublist(1).join(':').trim();
+        final newMap = <String, dynamic>{};
+        // If value is empty,
+        // create a new Map assuming a child element and add it to current.
+        if (value.isEmpty) {
+          current[key] = newMap;
+          newMap['__parent__'] = current;
+          stack.add(key);
+          current = newMap;
+        } else {
+          current[key] = value;
+        }
+      }
+    }
+    // Delete the "__parent__" key.
+    void cleanParent(Map<String, dynamic> map) {
+      map.remove('__parent__');
+      for (final value in map.values) {
+        if (value is Map<String, dynamic>) {
+          cleanParent(value);
+        }
+      }
+    }
+
+    cleanParent(result);
+    return Yaml._(result);
+  }
+
+  final YamlMap yamlMap;
+}
```
![nits-badge](https://img.shields.io/badge/review-nits-orange.svg) 
**Problem**
Currently, we store a __parent__ key in each nested map to track its parent. This approach can be confusing because __parent__ is not part of the actual data structure and must be removed afterward, risking potential oversights or collisions if the user’s YAML includes the same key.

**Reason**
By embedding a parent reference in the data, the final map is cluttered with an internal implementation detail. It also requires extra cleanup steps to remove the __parent__ key, which complicates maintenance and can lead to unexpected behavior if mishandled.

![ask-badge](https://img.shields.io/badge/review-ask-yellowgreen.svg) 
**Proposal**
We can manage nesting without a special key by using a stack of maps to represent the current nesting level. Whenever we indent further, we push a new map onto the stack; when we dedent, we pop from the stack. This removes the need for a __parent__ reference and keeps the resulting map clean and intuitive. For instance:

```dart
  factory Yaml.parse(String content) {
    final lines = content.split('\n');
    final result = <String, dynamic>{};

    // A stack of maps to manage the current nesting level.
    final mapStack = <Map<String, dynamic>>[result];

    for (final line in lines) {
      final trimmed = line.trim();
      if (trimmed.isEmpty || trimmed.startsWith('#')) {
        continue;
      }

      // Determine indentation by counting leading spaces.
      final indent = line.length - line.trimLeft().length;

      // If the stack's depth is deeper than the current indent, pop until it matches.
      while (mapStack.length > indent) {
        mapStack.removeLast();
      }

      final currentMap = mapStack.last;

      // Only process the line if it contains a colon.
      if (trimmed.contains(':')) {
        // Split the line into a key and value (joining the rest in case there are multiple colons).
        final parts = trimmed.split(':');
        final key = parts[0].trim();
        final value = parts.sublist(1).join(':').trim();

        // If the value is empty, treat the key as a nested map.
        if (value.isEmpty) {
          final newMap = <String, dynamic>{};
          currentMap[key] = newMap;
          mapStack.add(newMap);
        } else {
          // Otherwise, assign the value as a string.
          currentMap[key] = value;
        }
      }
    }

    return Yaml._(result);
  }
```

Would you consider applying this approach? It should simplify the structure by removing the internal __parent__ references and make the parser easier to maintain.

blendthink reviewed (commented):
@masa-futa 
> Changed Yaml.parse argument to string for easier testing.
Because of this, MemoryFileSystem is no longer necessary, so added support for removing it.

Thank you so much for your wonderful improvements!

I've left some comments on the pull request. Please take a look when you have a moment.

masa-futa commented:
@blendthink 
Thank you for pointing that out !!

🔸Delete unnecessary imports
⇨https://github.com/yumemi-inc/flutter-yumemi-lints/pull/149/commits/ddc183751964c9efa54cbec07d52556e6d766270

🔸Renaming variables
⇨https://github.com/yumemi-inc/flutter-yumemi-lints/pull/149/commits/f1818838d6ceef76433b385ff2f4d8a8fd058993

🔸Refactoring `Yaml.parse`
Thanks again for your cool and great suggestions !!
We ran the code you provided and found that some parts were not working correctly.  

*The contents use `currentMap` appropriately and changed to store child hierarchy values ​​hierarchically instead of in the same order as the parent hierarchy.
We fixed that part only and used the code you provided for the rest!
⇨https://github.com/yumemi-inc/flutter-yumemi-lints/pull/149/commits/fd3fdde5cafa29867e6869eb75eb3b7ff5e0eefc

We would appreciate it if you could check it again when you have a chance.

blendthink reviewed (approved):
@masa-futa 
Thank you for making a more appropriate correction! 
LGTM!
